lib/Parse/ParseIfConfig.cpp

```diff
@@ -623,8 +623,18 @@ ParserResult<IfConfigDecl> Parser::parseIfConfig(
     foundActive |= isActive;
 
     if (!Tok.isAtStartOfLine() && Tok.isNot(tok::eof)) {
-      diagnose(Tok.getLoc(),
-               diag::extra_tokens_conditional_compilation_directive);
+      if (isElse && Tok.is(tok::kw_if)) {
```

Comment:
I think you should do this before line 599. Like:

```diff
     bool isElse = Tok.is(tok::pound_else);
     SourceLoc ClauseLoc = consumeToken();
+    if (isElse && Tok.is(tok::kw_if) && !Tok.isAtStartOfLine()) {
+     isElse = false;
+      .. Emit error and fix-it ..
+    }
+
     Expr *Condition = nullptr;
     bool isActive = false;

     // Parse the condition.  Evaluate it to determine the active
     // clause unless we're doing a parse-only pass.
     if (isElse) {
```
This way, the parser automatically parse and possibly diagnose the condition after `#else if`.

For example, given
```swift
#else if SOME == THING
```
In this case, we want to diagnose both:
- `#else if` for fix-it `#elseif`
- `==` for `error: expected '&&' or '||' expression`

Reply:
Yeah, recovering here seems to work well and catch parse errors in the condition. Thanks!